pkg/ceph/osd/health.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2017 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package osd
+
+import (
+	"time"
+
+	"github.com/rook/rook/pkg/ceph/client"
+	"github.com/rook/rook/pkg/clusterd"
+)
+
+var (
+	healthCheckInterval = 60 * time.Second
+
+	// waitThreshold is period before an action is taken
+	waitThreshold = 300 * time.Second
+)
+
+// Monitor defines OSD process monitoring
+type Monitor struct {
+	context *clusterd.Context
+	agent   *OsdAgent
+
+	// lastStatus keeps track of OSDs status
+	// key - Process ID; value: time of the status change.
+	lastStatus map[int]time.Time
+}
+
+// NewMonitor instantiates OSD monitoring
+func NewMonitor(context *clusterd.Context, agent *OsdAgent) *Monitor {
+	return &Monitor{context, agent, make(map[int]time.Time)}
+}
+
+// Run runs monitoring logic for osd processe's status at set intervals
+func (m *Monitor) Run(stopCh chan struct{}) {
+	for {
+		select {
+		case <-stopCh:
+			logger.Infof("Stopping OSD monitoring.")
+			return
+
+		case <-time.After(healthCheckInterval):
+			logger.Debug("Checking osd processes status.")
+			err := m.procStatus()
+			if err != nil {
+				logger.Infof("Failed OSD status check: %s", err)
+			}
+		}
+	}
+}
+
+func (m *Monitor) procStatus() error {
+	osdDump, err := client.GetOSDDump(m.context, m.agent.cluster.Name)
+	if err != nil {
+		return err
+	}
+
+	// TODO: handling of OSDs that are not tracked by Agent or OSDDump
+	logger.Debugf("OSDDump vs. Agent.osdProc : %d", len(osdDump.OSDs)-len(m.agent.osdProc))
+
+	for id, proc := range m.agent.osdProc {
+		logger.Debugf("validating status of OSD id%d", id)
+		_, tracked := m.lastStatus[id]
+		// TODO: inCluster state action to take
```

Comment:
Not sure yet, how to handle out of cluster OSDs, as I understand there are valid reasons an OSD can be *out* such as data rebalancing.

Reply:
from the docs on removing an OSD and migrating its data, it looks like when an OSD is marked out by the user it is still seen as `in`.  in that case, you'd probably want the OSD to keep running in case it needs to assist in migrating its data.  Perhaps it's a good start for us to only be concerned with `down` but not do anything for `out` yet.